### ISSUE TITLE
Data races check strategy

### ISSUE THREAD

**Issue by XLiZHI** [esbmc/esbmc#1436]
```
#include <pthread.h>

void* t1(void *arg) {
    int* shared = (int*) arg;
    *shared = 1;                 

    return NULL;
}

int main() {
    pthread_t id1, id2;
    int shared_data;

    pthread_create(&id1, NULL, t1, &shared_data);

    return 0;
}
```

The following is the current principle of data races check, which can be roughly divided into four parts:

We use the identifier to create a new symbol: `"tmp_" + "c:main6.c@48@F@t1@shared"`

```
t1 (c:@F@t1):
        ...
        tmp_c:main6.c@48@F@t1@shared=1;  // (1). Set

        *shared=1;

        tmp_c:main6.c@48@F@t1@shared=0;  // (2). Reset

        ASSERT !tmp_c:main6.c@48@F@t1@shared // W/W data race on c:main6.c@48@F@t1@shared  // (3). Assertion

        RETURN: 0
        END_FUNCTION // t1
        
__ESBMC_main (__ESBMC_main):
       ...
        tmp_c:main6.c@48@F@t1@shared=0; // (4). Initialise 
```

Thus, a violation of assertion (3) during thread interleaving represents the occurrence of a data race.

The method for adding data race check is at the `goto` level, meaning that it is not thread-aware. Therefore, static analysis fails to dereference at the thread level.

But it would be a bit of a mistake to port all the processes to symex, which involves reachability trees and so on. I think the best way to do this is to implement it in goto functions and then dereferencing it in symex:

1. We add a special prefix (_ESBMC_deref) to symbols that need to be dereferenced.
2. In symex for the special prefix we dereferenced it.

It looks like this:
```
t1 (c:@F@t1):
        ...
        _ESBMC_deref_c:main6.c@48@F@t1@shared=1;  // (1). Split into "c:main6.c@48@F@t1@shared" and dereference
                                             to  get "shared_data", and create new symbol "tmp_c: shared_data"

        *shared=1;

        _ESBMC_deref_c:main6.c@48@F@t1@shared=0;  // (2). same to (1)

        ASSERT !_ESBMC_deref_c:main6.c@48@F@t1@shared // W/W data race on c:main6.c@48@F@t1@shared  // (3). Assertion

        RETURN: 0
        END_FUNCTION // t1
        
__ESBMC_main (__ESBMC_main):
       ...
        _ESBMC_deref_c:main6.c@48@F@t1@shared=0; // (4). Initialise 
```

The implementation is [here](https://github.com/XLiZHI/esbmc/blob/dba2545ba7db34b99f2a15df8bcc47b320a3e9c5/src/goto-symex/builtin_functions.cpp#L1511-L1565).

But there is a problem with this approach: we can dereferencing symbols in (1), (2), and (3). However, the position of (4) is in _ESBMC_main, and the corresponding value sets are not available in symex. This fails to initialize the temporary variable, causing assertion (3) to be violated.

**Comment by lucasccordeiro** [comment]
@XLiZHI: I see you're getting closer to the final solution. Can you please invest more time/effort in fixing the assertion (3) violation?

**Comment by XLiZHI** [comment]
> @XLiZHI: I see you're getting closer to the final solution. Can you please invest more time/effort in fixing the assertion (3) violation?

Hi @lucasccordeiro, sure, I opened this issue to ask for some advice. As shown above, we create the new symbol in symex, but we need to initialize it properly (**just like (4)**), but we can't derereference it in `_ESBMC_main` because we don't have the `shared`'s value set yet. 

I think our process at symex is like this:

①`_ESBMC_main()`  -> ②`function call：main()` -> ③`function call: pthread_create(..)` -> .... ->④`FUNCTION_CALL:  return_value$=*startdata.func(startdata.start_arg)`

In my investigation, we didn't get the `shared`’s value set until ④.

```
__ESBMC_main (__ESBMC_main):
       ...
        _ESBMC_deref_c:main6.c@48@F@t1@shared=0; // (4). Initialise
```

In summary, at this point we need to initialize a new symbol(tmp_c:shared_data), but we can't dereference it at this step. Maybe there are other ways we can do this initialization?

@fbrausse, May I have your opinion?

**Comment by XLiZHI** [comment]
close via #1460
